src/Processors/QueryPlan/ReadFromSystemNumbersStep.cpp

```diff
@@ -176,23 +176,31 @@ class NumbersRangedSource : public ISource
     {
         std::lock_guard lock(ranges_state->mutex);
 
-
         UInt64 need = base_block_size_;
+        bool without_block_size_limit = need == 0;
```

Comment:
I think the code would be simpler if we handled this case outside of the loop.

That being said, I'm unsure about what `max_block_size = 0` means. Is it return everything in one block (scary) or should it be something else (like use the default)? Maybe we should handle this case via setting quirks / constrains internally to define it.

Reply:
yeah, I agree with all your points
I wanted to fix just https://github.com/ClickHouse/ClickHouse/pull/63587/files#diff-2d15d3775e0963c3015ee4f657d6712d8dcaeddd5a8be535947df446eb092f83R538 and avoid division by 0 then I noticed that we return 0 instead of correct result.

I did quick grep to see how we handle those cases in general and only found this https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/KVStorageUtils.cpp#L270
where we return everything in a single block.

But yeah, it could be much scarier for an infinite source like system.numbers.

Reply:
In the short term, maybe we can extend the `doSettingsSanityCheckClamp` function to turn 0 into the default value or some other value that makes sense (maybe 1)?

Reply:
That seems a bit too confusing.
Maybe simply throw an exception if setting like `max_block_size` is set to 0?

Reply:
If you do this, somebody will cry because of backward incompatibility and upgrades and queries will break (talking from experience). In the near short time I plan to add some options about sanity checks to select whether you want to clamp or throw, but for now we can just ignore it and set it to something that's ok

Reply:
But we will break compatibility in any case if the sources behave differently depending on the value of `max_block_size`.

To avoid being stuck on this, I will set it to default inside `doSettingsSanityCheckClamp`

Reply:
I've checked how we handle it in MergeTree and found this
https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/QueryPlan/ReadFromMergeTree.cpp#L410
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeReadTask.cpp#L159

We will change the behavior but in any case it shouldn't be noticeable from user